res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -504,3 +479,143 @@ Prime4.shiftState = function(channel, control, value) {
         Prime4.rightDeck.unshift();
     }
 };
+
+//========== PERFORMANCE PADS ==========//
+
+Prime4.padMode = {
+    HOTCUE: 0x0B,
+    LOOP: 0x0C,
+    ROLL: 0x0D,
+    SLICER: 0x0E,
+};
+
+Prime4.padSection = function(deck, offset) {
+    components.ComponentContainer.call(this);
+    this.modes = {
+        "hotcue": new Prime4.hotcueMode(deck, offset),
+        // "loop": new Prime4.loopMode(deck, offset),
+        // "autoloop": new Prime4.autoloopMode(deck, offset),
+        "roll": new Prime4.rollMode(deck, offset),
+        // "sampler": new Prime4.samplerMode(deck, offset),
+    };
+    this.offset = offset;
+
+    // Start in Hotcue mode
+    this.setPadMode(Prime4.padMode.HOTCUE);
+    //midi.sendShortMsg(0x94 + offset, 0x0C, Prime4.rgbCode.whiteDark); // Loop mode, not implemented yet
+    midi.sendShortMsg(0x94 + offset, 0x0D, Prime4.rgbCode.whiteDark);
+    //midi.sendShortMsg(0x94 + offset, 0x0E, Prime4.rgbCode.whiteDark); // Slicer mode, not implemented yet
+};
+
+Prime4.padSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+Prime4.padSection.prototype.controlToPadMode = function(control) {
+    let mode;
+    switch (control) {
+    case Prime4.padMode.HOTCUE:
+        mode = this.modes.hotcue;
+        break;
+    case Prime4.padMode.LOOP:
+        if (this.currentMode === this.modes.loop) {
+            mode = this.modes.autoloop;
+        } else {
+            mode = this.modes.loop;
+        }
+        break;
+    case Prime4.padMode.ROLL:
+        mode = this.modes.roll;
+        break;
+    case Prime4.padMode.SLICER:
+        mode = this.modes.sampler;
+        break;
+    }
+
+    return mode;
+};
+
+Prime4.padSection.prototype.padModeButtonPressed = function(channel, control, value, _status, _group) {
+    if (value) {
+        this.setPadMode(control);
+    }
+};
+
+// Worry about parameter buttons later
+//Prime4.padSection.prototype.paramButtonPressed = function(channel, control, value, status, group) {};
+
+Prime4.padSection.prototype.setPadMode = function(control) {
```

Comment:
instead of specifing these on the prototype like that, you can also specify them in the `padSection` constructor via `this.setPadMode = function(...){...}`. That has the added benefit that you can refer to arguments given to that constructor.

Reply:
Sweet. I just followed the structure of the DJ505 mapping, but any way I can improve it is greatly appreciated

Reply:
Well its really a style choice, either is fine. I just wanted to give you that hint in case you need access the constructor parameter (or other local variables in the constructor) for something you want to implement.

Reply:
Thank you